### ISSUE TITLE
Internal `__stream_shard__` label is visible to user

### ISSUE THREAD

**Issue by z0rc** [grafana/loki#13095]
**Describe the bug**
https://grafana.com/docs/loki/latest/operations/automatic-stream-sharding/ is enabled by default. It creates label `__stream_shard__` that is visible to user when searching.

**To Reproduce**
1. Loki 3.0.0, Grafana v10.4.1
2. Open Grafana Explore, start building query
3. Observe `__stream_shard__` being offered in label suggestions. Also it's present in search result fields.

**Expected behavior**
`__stream_shard__` shouldn't be visible to user, it provides zero value when searching.

**Environment:**
 - Infrastructure: Kubernetes
 - Deployment tool: helm

**Screenshots, Promtail config, or terminal output**
<img width="540" alt="Screenshot 2024-05-31 at 16 50 37" src="https://github.com/grafana/loki/assets/787519/20c3b0dd-7454-4aa6-83ee-39c56aad546b">

<img width="1603" alt="Screenshot 2024-06-01 at 16 52 26" src="https://github.com/grafana/loki/assets/787519/82aa73b5-b03e-4921-b9a8-cff21e054c53">

**Comment by DylanGuedes** [comment]
We weren't strongly opinionated about keeping the label, but having it can be useful. Example:
- If you would like to know how big a given stream is
- If you would like to reduce your search space to have a faster query
- Hiding it makes the feature more obscure

**Comment by z0rc** [comment]
> If you would like to know how big a given stream is

Stream shard count (number?) is kinda obscure metric for this. Can it be represented in proper prom metric?

> If you would like to reduce your search space to have a faster query

For me it reads not as "faster", but "incomplete" at best, "incorrect" at worst. Isn't there better options to make search faster, like bloom filters?

> Hiding it makes the feature more obscure

But keeping it places makes this visually inconsistent and overloaded with information that isn't needed.

[Promatail documentation](https://grafana.com/docs/loki/latest/send-data/promtail/scraping/#file-target-discovery) has this:

> Labels starting with __ (two underscores) are internal labels. They usually come from dynamic sources like service discovery. Once relabeling is done, they are removed from the label set. To persist internal labels so they’re sent to Grafana Loki, rename them so they don’t start with __.

Which I think applicable to `__stream_shard__` too.

I understand your position from development perspective, but from end user I still consider this label as noise, which I what to reduce.

**Comment by DylanGuedes** [comment]
> > If you would like to know how big a given stream is
> 
> Stream shard count (number?) is kinda obscure metric for this. Can it be represented in proper prom metric?

Hardly, because it would have infinite cardinality.

> > If you would like to reduce your search space to have a faster query
> 
> For me it reads not as "faster", but "incomplete" at best, "incorrect" at worst. Isn't there better options to make search faster, like bloom filters?

Yes. I wasn't saying it was an ideal solution for that use case; but yeah, I've used it myself a few times.

> 
> > Hiding it makes the feature more obscure
> 
> But keeping it places makes this visually inconsistent and overloaded with information that isn't needed.
> 
> [Promtail documentation](https://grafana.com/docs/loki/latest/send-data/promtail/scraping/#file-target-discovery) has this:
> 
> > Labels starting with __ (two underscores) are internal labels. They usually come from dynamic sources like service discovery. Once relabeling is done, they are removed from the label set. To persist internal labels so they’re sent to Grafana Loki, rename them so they don’t start with __.
> 
> Which I think applicable to `__stream_shard__` too.
> 
> I understand your position from development perspective, but from end user I still consider this label as noise, which I want to reduce.

That is super fair, thanks for bringing it. As I said in my previous message, we weren't strongly opinionated about keeping it in the past, but we just left it as we didn't find strong opposition either. Now that 3.0 is released with this enabled by default we might see more and more users against it though. I'll add to my backlog to revive the discussion and maybe hide it in the near future.

**Comment by Rohlik** [comment]
I agree with @z0rc, as our end users are also confused by this label. I vote for hiding it.

**Comment by jtackaberry** [comment]
I would also like to hide `__tenant_id__` for the same reason.  We have a number of multi-tenant data sources, but [`__tenant_id__` isn't actually a useful label for querying](https://github.com/grafana/loki/issues/13259) and, as others have said about `__stream_shard__`, it's just noise.

Absent excluding on the Loki side, I'd be happy if Grafana itself just filtered this label out when presenting to the users.  I'm less fussed if the labels is surfaced to CLI and API users as these users tend to be less easy to confuse.

**Comment by IntegersOfK** [comment]
> I would also like to hide `__tenant_id__` for the same reason. We have a number of multi-tenant data sources, but [`__tenant_id__` isn't actually that useful a label for querying](https://github.com/grafana/loki/issues/13259) and, as others have said about `__stream_shard__`, it's just noise.

The `__tenant_id__` label is the only one that allows us to discriminate between multiple tenants when using a data source that is configured with multiple tenants, which is a valid way to use Loki (it allows pipe separation in the X-Scope-OrgID header value).

I would suggest you keep the scope of this issue to only the other non-useful internal labels. For example, my organization relies on `__tenant_id__` for limiting queries to the scope of a specific client's telemetry.

**Comment by jtackaberrybb** [comment]
> The issue you mentioned about that, is a bug, not the intended behaviour and it's not how every other version has worked since multiple tenants were enabled. I wouldn't use that as rationale for the statement that it "isn't actually that useful."

Well fair enough. `__tenant_id__` could and should be useful. Though personally I wish there was a way to alias `__tenant_id__` to something that doesn't scream "this is internal and you shouldn't use this."  At least that's certainly how any of our Python developers would interpret it :)

**Comment by jtackaberry** [comment]
> The issue you mentioned about that, is a bug, not the intended behaviour and it's not how every other version has worked since multiple tenants were enabled. I wouldn't use that as rationale for the statement that it "isn't actually that useful."

Well fair enough. __tenant_id__ could and should be useful. Though personally I wish there was a way to alias __tenant_id__ to something that doesn't scream "this is internal and you shouldn't use this." At least that's certainly how any of our Python developers would interpret it :)

**Comment by IntegersOfK** [comment]
Sure yes alias would be nice, or a config/parameter to include/exclude that injection behaviour would be acceptable too in my opinion.

**Comment by PRAD-KANAKAPURA** [comment]
Hello, I have a strange behavior after upgrading Grafana and Loki from v10.4 to v11.2 and v2.9.2 to v3.2.0. The dashboard now have duplicate values but the later one having the extra label names `__stream_shard__`. For other panels I have no problem. Could anyone guide through this pls?
![Image](https://github.com/user-attachments/assets/cc2eea3d-dff4-462e-b3f1-d9499e592406)

**Comment by vladst3f** [comment]
@DylanGuedes , can you hide this __stream_shard__ label, pretty please ? I believe there's a consesus that this is useless for querying and it's so confusing for users that I bet some installations have sharding off just to avoid this noise. Cheers!

**Comment by svennergr** [comment]
Closed by https://github.com/grafana/grafana/pull/97323 in Grafana.
